eZ/Publish/Core/IO/IOMetadataHandler/LegacyDFSCluster.php

```diff
@@ -59,30 +59,39 @@ public function create(SPIBinaryFileCreateStruct $binaryFileCreateStruct)
         }
 
         $path = (string)$this->addPrefix($binaryFileCreateStruct->id);
-
-        try {
-            /*
-             * @todo what might go wrong here ? Can another process be trying to insert the same image ?
-             *       what happens if somebody did ?
-             **/
-            $stmt = $this->db->prepare(
-                <<<SQL
+        $params = [
+            'name' => $path,
+            'name_hash' => md5($path),
+            'name_trunk' => $this->getNameTrunk($binaryFileCreateStruct),
+            'mtime' => $binaryFileCreateStruct->mtime->getTimestamp(),
+            'size' => $binaryFileCreateStruct->size,
+            'scope' => $this->getScope($binaryFileCreateStruct),
+            'datatype' => $binaryFileCreateStruct->mimeType,
+        ];
+
+        if ($this->db->getDatabasePlatform()->getName() === 'postgresql') {
+            $sql = <<<SQL
+INSERT INTO ezdfsfile
+  (name, name_hash, name_trunk, mtime, size, scope, datatype)
+  VALUES (:name, :name_hash, :name_trunk, :mtime, :size, :scope, :datatype)
+ON CONFLICT (name_hash) DO UPDATE
+  SET datatype=:datatype, scope=:scope, size=:size, mtime=:mtime
+  WHERE ezdfsfile.name_hash=:name_hash
+SQL;
+        } else {
+            $sql = <<<SQL
 INSERT INTO ezdfsfile
   (name, name_hash, name_trunk, mtime, size, scope, datatype)
   VALUES (:name, :name_hash, :name_trunk, :mtime, :size, :scope, :datatype)
 ON DUPLICATE KEY UPDATE
   datatype=VALUES(datatype), scope=VALUES(scope), size=VALUES(size),
   mtime=VALUES(mtime)
-SQL
-            );
-            $stmt->bindValue('name', $path);
-            $stmt->bindValue('name_hash', md5($path));
-            $stmt->bindValue('name_trunk', $this->getNameTrunk($binaryFileCreateStruct));
-            $stmt->bindValue('mtime', $binaryFileCreateStruct->mtime->getTimestamp());
-            $stmt->bindValue('size', $binaryFileCreateStruct->size);
-            $stmt->bindValue('scope', $this->getScope($binaryFileCreateStruct));
-            $stmt->bindValue('datatype', $binaryFileCreateStruct->mimeType);
-            $stmt->execute();
+SQL;
+        }
+
+
+        try {
+            $this->db->executeStatement($sql, $params);
```

Comment:
This is not a clean way to do it. It either has to be uniform ANSI SQL (can be more than one statement) or a strategy depending on DB. ATM this violates Open-Closed principle of SOLID.